ts/ui/safe/safe.ts

```diff
@@ -0,0 +1,267 @@
+    //    Otherwise use [-lengthMax,lengthMax] from above
+    //
+    styleLengths: expandable({
+      borderTop: 'borderTopWidth',
+      borderRight: 'borderRightWidth',
+      borderBottom: 'borderBottomWidth',
+      borderLeft: 'borderLeftWidth',
+      paddingTop: true,
+      paddingRight: true,
+      paddingBottom: true,
+      paddingLeft: true,
+      marginTop: true,
+      marginRight: true,
+      marginBottom: true,
+      marginLeft: true,
+      outlineTop: true,
+      outlineRight: true,
+      outlineBottom: true,
+      outlineLeft: true,
+      fontSize: [.707, 1.44]
+    })
+  };
+
+  /**
+   * The attribute-to-filter-method mapping
+   */
+  public filterAttributes: {[name: string]: string} = {
+    //
+    //  Methods called for MathML attribute processing
+    //
+    href:                 'filterURL',
+    src:                  'filterURL',
+    altimg:               'filterURL',
+    'class':              'filterClassList',
+    style:                'filterStyles',
+    id:                   'filterID',
+    fontsize:             'filterFontSize',
+    mathsize:             'filterFontSize',
+    scriptminsize:        'filterFontSize',
+    scriptsizemultiplier: 'filterSizeMultiplier',
+    scriptlevel:          'filterScriptLevel',
+    'data-':              'filterData'
+  };
+
+  /**
+   * The safe options from the document option list
+   */
+  public options: OptionList;
+
+  /**
+   * Shorthand for options.allow
+   */
+  public allow: OptionList;
+
+  /**
+   * The DOM adaptor from the document
+   */
+  public adaptor: DOMAdaptor<N, T, D>;
+
+  /**
+   * The methods for filtering the MathML attributes
+   */
+  public filterMethods: {[name: string]: FilterFunction<N, T, D>} = {
+    ...SafeMethods
+  };
+
+  /**
+   * @param {MathDocument<N,T,D>} document  The MathDocument we are sanitizing
+   * @param {OptionList} options            The safeOptions from the document
+   */
+  constructor(document: MathDocument<N, T, D>, options: OptionList) {
+    this.adaptor = document.adaptor;
+    this.options = options;
+    this.allow = this.options.allow;
+  }
+
+  /**
+   * Sanitize a MathItem's root MathML tree
+   *
+   * @param {MathItem<N,T,D>} math           The MathItem to sanitize
+   * @param {MathDocument<N,T,D>} document   The MathDocument in which it lives
+   */
+  public sanitize(math: MathItem<N, T, D>, document: MathDocument<N, T, D>) {
+    try {
+      math.root.walkTree(this.sanitizeNode.bind(this));
+    } catch (err) {
+      document.options.compileError(document, math, err);
+    }
+  }
+
+  /**
+   * Sanitize a node's attributes
+   *
+   * @param {MmlNode} node      The node to sanitize
+   */
+  protected sanitizeNode(node: MmlNode) {
+    const attributes = node.attributes.getAllAttributes();
+    for (const id of Object.keys(attributes)) {
+      if (this.filterAttributes.hasOwnProperty(id)) {
+        const value = this.filterMethods[this.filterAttributes[id]](this, attributes[id]);
```

Comment:
Is it certain that for every `filterAttribute` there will be a `filterMethod`?

Reply:
Nothing is certain.  What are you suggesting?  Checking and reporting an error if not?  Isn't that what will happen anyway if not?  Because this will be called on every attribute on every node, it should be kept as slim as possible.  If someone adds a `filterAttribute` without creating the corresponding `filterMethod`, then they will get an error message at this point.

Reply:
I was wondering if we should catch that error at this point, i.e., if the `filterMethod` does not exist, then we could simply continue in the loop, or throw an explicit error, rather than get a `undefined is not a function` error.